vdom/core.py

```diff
@@ -56,157 +52,95 @@ def to_json(el, schema=None):
 
     return json_el
 
-
 class VDOM(object):
     """A basic virtual DOM class which allows you to write literal VDOM spec
 
-    >>> VDOM({ 'tagName': 'h1', 'children': 'Hey', 'attributes': {}})
-
-    It's probably better to use `vdom.h` or the helper functions:
-
-    >>> from vdom import h
-    >>> h('h1', 'Hey')
-    <h1 />
+    >>> VDOM(tag_name='h1', children='Hey', attributes: {}})
 
     >>> from vdom.helpers import h1
     >>> h1('Hey')
-
     """
-    _schema = VDOM_SCHEMA
-    _obj = None
+    # This class should only have these 4 attributes
+    __slots__ = ['tag_name', 'attributes', 'children', 'key']
 
-    def __init__(self, obj, schema=None):
-        # we need to assign self.schema first,
-        # because it is used to validate the object
+    def __init__(self, tag_name, attributes=None, children=None, key=None, schema=None):
         if schema is not None:
-            self.schema = schema
-        self.obj = obj
+            warnings.warn('Passing schema param to VDOM constructor is deprecated')
+        if isinstance(tag_name, dict) or isinstance(tag_name, list):
+            # Backwards compatible interface
+            warnings.warn('Passing dict to VDOM constructor is deprecated')
+            value = tag_name
+            vdom_obj = VDOM.from_dict(value)
+            tag_name = vdom_obj.tag_name
+            attributes = vdom_obj.attributes
+            children = vdom_obj.children
+            key = vdom_obj.key
+        self.tag_name = tag_name
+        self.attributes = attributes if attributes else {}
+        self.children = children if children else []
+        self.key = key
+
+    def to_dict(self):
+        vdom_dict = {
+            'tagName': self.tag_name,
+            'attributes': self.attributes
+        }
+        if self.key:
+            vdom_dict['key'] = self.key
```

Comment:
I almost wish that we always had a key… that would make the later custom schema stuff much easier to think about since it would be the same kind of object all the way down… 

Even if the default key was something like `'VDOM-component-list'` that'd be a nice way to convey what it is at the top level rather than having no key there at all.

Reply:
I agree! but I wanted to be as backwards compatible as possible, and limit the number of changes I am making. Maybe in a different PR.

Reply:
We shouldn't _always_ have a key as it gets for indicating how a (real) virtual dom what later versions should replace which nodes in a tree. This conditional checking is wise.

Reply:
Unless of course you're saying it would always be at least `None`. I can't recall if we declared in the spec that it was nullable (instead of optional). I think it's better to send no bytes than an explicit `null` for the key.